crates/rattler_package_streaming/src/tokio/async_read.rs

```diff
@@ -3,32 +3,156 @@
+    reader: impl AsyncRead + Send + Unpin + 'static,
     destination: &Path,
 ) -> Result<ExtractResult, ExtractError> {
-    // Create a async -> sync bridge
-    let reader = SyncIoBridge::new(Box::pin(reader));
+    use async_compression::tokio::bufread::BzDecoder;
 
-    // Spawn a block task to perform the extraction
+    // Ensure the destination directory exists
+    tokio::fs::create_dir_all(destination)
+        .await
+        .map_err(ExtractError::CouldNotCreateDestination)?;
+
+    // Clone destination for the async block
     let destination = destination.to_owned();
-    match tokio::task::spawn_blocking(move || crate::read::extract_tar_bz2(reader, &destination))
+
+    // Wrap the reading in additional readers that will compute the hashes while extracting
+    let sha256_reader = rattler_digest::HashingReader::<_, rattler_digest::Sha256>::new(reader);
+    let mut md5_reader =
+        rattler_digest::HashingReader::<_, rattler_digest::Md5>::new(sha256_reader);
+    let mut size_reader = SizeCountingReader::new(&mut md5_reader);
+
+    // Create a buffered reader for better performance
+    let buf_reader = tokio::io::BufReader::with_capacity(DEFAULT_BUF_SIZE, &mut size_reader);
+
+    // Decompress bzip2 asynchronously
+    let decoder = BzDecoder::new(buf_reader);
+
+    // Build archive with optimized settings for faster extraction:
+    // - Skip mtime preservation to avoid extra syscalls
+    // - Skip automatic permission handling (we'll set executable bits manually)
+    // - Skip extended attributes for better performance
+    let archive = tokio_tar::ArchiveBuilder::new(decoder)
+        .set_preserve_mtime(false)
+        .set_preserve_permissions(false)
+        .set_unpack_xattrs(false)
+        .build();
+
+    // Unpack entries manually, preserving only executable bits on Unix
+    unpack_tar_archive(archive, &destination).await?;
+
+    // Read the file to the end to make sure the hash is properly computed
+    tokio::io::copy(&mut size_reader, &mut tokio::io::sink())
         .await
-    {
-        Ok(result) => result,
-        Err(err) => {
-            if let Ok(reason) = err.try_into_panic() {
-                std::panic::resume_unwind(reason);
-            }
-            Err(ExtractError::Cancelled)
-        }
+        .map_err(ExtractError::IoError)?;
+
+    // Get the size and hashes
+    let (_, total_size) = size_reader.finalize();
+    let (sha256_reader, md5) = md5_reader.finalize();
+    let (_, sha256) = sha256_reader.finalize();
+
+    // Validate that we actually read some data from the stream.
+    // If total_size is 0, it likely means the stream was truncated or the bzip2
+    // decompressor silently failed without detecting an incomplete stream.
+    if total_size == 0 {
+        return Err(ExtractError::IoError(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            "no data was read from the package stream - the stream may have been truncated",
+        )));
     }
```

Comment:
We have to do this because of this issue https://github.com/Nullus157/async-compression/issues/411 .

I made an upstream fix in https://github.com/Nullus157/async-compression/pull/412. The release is till pending though: https://github.com/Nullus157/async-compression/pull/393